src/Meziantou.Framework.WPF/Collections/Internals/DispatchedObservableCollection.cs

```diff
@@ -206,54 +232,112 @@ private void EnqueueEvent(PendingEvent<T> @event)
+            foreach (var @event in events)
+            {
+                _pendingEvents.Enqueue(@event);
+            }
+
+            ProcessPendingEventsOrDispatch();
+        }
+
+        private void ProcessPendingEventsOrDispatch(bool forceProcess = false)
+        {
+            if (!forceProcess && _batchMode != null)
+                return;
+
             if (!IsOnDispatcherThread())
             {
                 if (!_isDispatcherPending)
                 {
                     _isDispatcherPending = true;
-                    _dispatcher.BeginInvoke((Action)ProcessPendingEvents);
+                    var mode = _batchMode;
+                    _dispatcher.BeginInvoke(new Action(() => ProcessPendingEvents(mode)));
                 }
 
                 return;
             }
 
-            ProcessPendingEvents();
+            ProcessPendingEvents(_batchMode);
         }
 
-        private void ProcessPendingEvents()
+        private void ProcessPendingEvents(BatchMode? batchMode)
         {
             _isDispatcherPending = false;
-            while (_pendingEvents.TryDequeue(out var pendingEvent))
+            var raiseEvents = batchMode != BatchMode.Reset;
+            foreach (var pendingEvent in DequeueEvents())
             {
                 switch (pendingEvent.Type)
                 {
                     case PendingEventType.Add:
-                        AddItem(pendingEvent.Item);
+                        AddItem(pendingEvent.Item, raiseEvents);
                         break;
 
                     case PendingEventType.Remove:
-                        RemoveItem(pendingEvent.Item);
+                        RemoveItem(pendingEvent.Item, raiseEvents);
                         break;
 
                     case PendingEventType.Clear:
-                        ClearItems();
+                        ClearItems(raiseEvents);
                         break;
 
                     case PendingEventType.Insert:
-                        InsertItem(pendingEvent.Index, pendingEvent.Item);
+                        InsertItem(pendingEvent.Index, pendingEvent.Item, raiseEvents);
                         break;
 
                     case PendingEventType.RemoveAt:
-                        RemoveItemAt(pendingEvent.Index);
+                        RemoveItemAt(pendingEvent.Index, raiseEvents);
                         break;
 
                     case PendingEventType.Replace:
-                        ReplaceItem(pendingEvent.Index, pendingEvent.Item);
+                        ReplaceItem(pendingEvent.Index, pendingEvent.Item, raiseEvents);
                         break;
                 }
             }
+
+            if (batchMode == BatchMode.Reset)
+            {
+                RaiseResetEvent();
+            }
+        }
+
+        private IEnumerable<PendingEvent<T>> DequeueEvents()
+        {
+            if (_batchMode == BatchMode.Optimized)
+            {
+                var events = new List<PendingEvent<T>>();
```

Comment:
I'm not sure the Optimized flow works well with WPF. I think we should make more tests.

Reply:
I've looked at the code of the WPF controls. It seems the old value and index are important.
So, I think this is only possible to combine events that could be replaced by a Replace event or no event. For instance:

- Remove item, then insert a new item at this index => Replace the old value with the new one
- Replace the same item twice => One replace with the original value and the new value
- Add Last / Remove last => No event
- ...

Note: If you think this is too complex or don't want to do that, it's ok to only keep `BatchMode.Reset` in this PR. Then, you (or someone else) can create another PR to implement the optimized mode.